src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,53 +12,57 @@ function calc_error_norms(func, u, t, analyzer,
   # TODO: MPI, we should improve this; maybe we should dispatch on `u`
   #       and create some MPI array type, overloading broadcasting and mapreduce etc.
   #       Then, this specific array type should also work well with DiffEq etc.
-  l2_error, linf_error = invoke(calc_error_norms,
+  l2_errors, linf_errors = invoke(calc_error_norms_per_element,
     Tuple{typeof(func), typeof(u), typeof(t), typeof(analyzer), TreeMesh{2},
           typeof(equations), typeof(initial_condition), typeof(dg), typeof(cache),
           typeof(cache_analysis)},
     func, u, t, analyzer, mesh, equations, initial_condition, dg, cache, cache_analysis)
 
-  # Since the local L2 norm is already normalized and square-rooted, we need to undo this first
-  total_volume = mesh.tree.length_level_0^ndims(mesh)
-  global_l2_error = Vector(l2_error.^2 .* total_volume)
-  global_linf_error = Vector(linf_error)
-  MPI.Reduce!(global_l2_error, +, mpi_root(), mpi_comm())
-  MPI.Reduce!(global_linf_error, max, mpi_root(), mpi_comm())
+  nvars = length(l2_errors[1])
+  T = typeof(l2_errors[1]) # for type conversion of final errors
+
+  # Convert errors from Vector of SVectors to Matrix for MPI communication
+  FT = eltype(l2_errors[1])
+  l2_errors = reshape(reinterpret(FT, l2_errors), nvars, :) |> Matrix
+  linf_errors = reshape(reinterpret(FT, linf_errors), nvars, :) |> Matrix
+
+  # Collect local error norms of all elements on root process
   if mpi_isroot()
-    l2_error   = convert(typeof(l2_error),   global_l2_error)
-    linf_error = convert(typeof(linf_error), global_linf_error)
+    global_l2_errors = zeros(eltype(l2_errors), nvars, cache.mpi_cache.n_elements_global)
+    global_linf_errors = similar(global_l2_errors)
+
+    n_elements_by_rank = parent(cache.mpi_cache.n_elements_by_rank) # convert OffsetArray to Array
+    l2_buf = MPI.VBuffer(global_l2_errors, nvars*n_elements_by_rank)
+    linf_buf = MPI.VBuffer(global_linf_errors, nvars*n_elements_by_rank)
+    MPI.Gatherv!(l2_errors, l2_buf, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, linf_buf, mpi_root(), mpi_comm())
   else
-    l2_error   = convert(typeof(l2_error),   NaN * global_l2_error)
-    linf_error = convert(typeof(linf_error), NaN * global_linf_error)
+    MPI.Gatherv!(l2_errors, nothing, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, nothing, mpi_root(), mpi_comm())
   end
 
-  l2_error = @. sqrt(l2_error / total_volume)
+  # Aggregate element error norms on root process
+  if mpi_isroot()
+    # Convert from Matrix to Vector of SVectors
+    global_l2_errors = reinterpret(T, vec(global_l2_errors))
+    global_linf_errors = reinterpret(T, vec(global_linf_errors))
 
-  return l2_error, linf_error
-end
+    # Aggregate element errors
 
+    # global_l2_errors is a Base.ReinterpretArray, by accessing it via [:] we convert it into a
+    # proper Vector so that sum always produces the same result as in the serial case
+    l2_error = sum(global_l2_errors[:])
```

Comment:
This allocates. Maybe in this specific case it would be easier to just use a hand-written sum loop for both the serial and the parallel case, such that you know that the same algorithm will be used exactly?

Reply:
Implemented a hand-written loop in 935e0de.

Reply:
:+1: